pinot-server/src/main/java/org/apache/pinot/server/starter/helix/HelixInstanceDataManager.java

```diff
@@ -242,6 +242,30 @@ public void addRealtimeSegment(String realtimeTableName, String segmentName)
     LOGGER.info("Added segment: {} to table: {}", segmentName, realtimeTableName);
   }
 
+  private void snapshotAndAddSegments(TableDataManager tableDataManager, String realtimeTableName)
+      throws Exception {
+    // we need to do an atomic snapshot here and then add segments because
```

Comment:
This one has race condition because when we create the new table data manager, there can be other operations applied to the old table data manager (add/remove/replace segment) which are not reflected in the new table data manager. We need to lock up the old table data manager until it is fully replaced

Reply:
That particular race condition is avoided because this logic is put inside _tableDataManagerMap.compute method which ends up locking the corresponding hashNode until the logic completes.

Have seen similar patterns in other functions of the code -- so I maintained the same.

I'll try to clarify it in an in-code comment -- let me know if I'm missing something.

Reply:
Oh, didn't notice that this is within the `computeIfPresent()`. So the race condition is not there, but it will block all the queries at the same time. In order to solve this, we still need to use lock so that queries can still acquire the old table data manager

Reply:
The `getTableDataManager` method uses `get` which doesn't lock the underlying hash-node. The add/delete/reload methods do a compute* on the hashmap which would invoke lock. So this locking wouldn't happen - the only case that remains is what you called out below about existing queries in the old table data manager. So I'll add a delayed shutdown there.

Let me know if I'm missing something.